README.md
# SimulIDE 0.3.10

Electronic Circuit Simulator


SimulIDE is a simple real time electronic circuit simulator.

It's intended for general purpose electronics and microcontroller simulation, supporting PIC, AVR and Arduino.

PIC simulation is provided by gpsim and avr simulation by simavr.

This is not an accurate simulator for circuit analisis, it aims to be the fast, simple and easy to use, so this means simple and not very accurate electronic models and limited features.

Intended for hobbist or students to learn and experiment with simple circuits.


SimulIDE also features a code Editor and Debugger for GcBasic, Arduino, PIC asm and AVR asm.
Editor/Debugger is still in it's firsts stages of development, with basic functionalities, but it is possible to write, compile and basic debugging with breakpoints, watch registers and global variables.


## Building SimulIDE:

Build dependencies:

 - Qt5 dev packages
 - Qt5Core
 - Qt5Gui
 - Qt5Xml
 - Qt5Widgets
 - Qt5Concurrent
 - Qt5svg dev
 - Qt5 Multimedia dev
 - Qt5 Serialport dev
 - Qt5 Script
 - Qt5 qmake
 - libelf dev
 - gcc-avr
 - avr-libc

 
Once installed go to build_XX folder, then:

```
$ qmake
$ make
```

In folder build_XX/release/SimulIDE_x.x.x you will find executable and all files needed to run SimulIDE.



## Running SimulIDE:

Run time dependencies:

 - Qt5Core
 - Qt5Gui
 - Qt5Xml
 - Qt5svg
 - Qt5Widgets
 - Qt5Concurrent
 - Qt5 Multimedia
 - Qt5 Multimedia Plugins
 - Qt5 Serialport
 - Qt5 Script
 - libelf


SimuliDE executable is in bin folder.
No need for installation, place SimulIDE folder wherever you want and run the executable.

## Compiling macOS

you need to have Qt and `libelf` installed via HomeBrew

    brew install qt libelf

and the inside the `build_XX` folder run

    /System/Volumes/Data/usr/local/Cellar/qt/5.14.1/bin/qmake
    macdeployqt ./release/SimulIDE_0.3.12-SR8/bin/simulide.app -dmg




changelog.txt

simulide 0.3.12


New Features:
  RC1:
    - Show Circuit Time and Mcu freq.
    - Show Led polarity.
    - Show Meters polarity.
    - MCU freq. decimal values.
    - Chip/Logic_Symbol for subcircuits.

  RC2:
    - Bjt BC diode.
    - Bjt faster switch.
    - multi-channel plotter, resizable scale.
    - Wave gen. frequency decimal values.
    - Oscope improved.
    - Windows OS console output.
    - Remove selected connectors.
    - Auto-connect pins.
    - Add short info to Component names.
    - Refactoring Bus: Fully analogic bidirectional Buses.
    
  RC3:
    - Bus start Bit.
    - Debugger GcBasic Step Over (experimental).
    - File Browser: Show hidden.
    - File Browser: Settings bookmark.
    - Graphic Package creation.
    - Ramtable case insensitive.
    - 74 IC Logic Symbols.
  
  SR1:
    - CD IC Logic Symbols.
    - Help widget.
    - Some help files.
    - Ramtable: 4 cols: Name, Type, Dec.Value, Value.
    - GcBasic: Strings in RamTable
    - Smaller Gates.
    - MCU EEPROM persistence (to .simu File).
    - Load/Save MCU EEPROM Data to data file.
    - Add persistence to RAM = PROM (to .simu File).
    - Resizable RAM/PROM ( address bits and word size ).
    - Load/Save RAM/PROM Data to data file.
    - Add persistence to I2C RAM = I2C PROM (to .simu File).
    - Load/Save I2C RAM/PROM Data to data file.
    - Auto-Backup and Restore Circuit after Crash.
    - Property to set Auto-Backup check interval( Secs ).
    
  SR2:
  
  SR3:
    - Led shape to original, but displaced to Cathode.
    *- 16 bit Encoder/Decoder
    
  SR4:
    - Circuit Zoom with key sequence: Ctrl+ Ctrl-
    - Serial Terminal: "Clear" and "CR" buttons.
    - New Dialog: create and edit Package Pins.
    - Create Subcircuit and Package files in one shot.
  
Bug Fixes:
  RC1:
    - Avoid closing document while debugging.
    - Modification in circuit while debugging or simulating MCUs.
    - Atmega8 not working.
    - LM555 power pins.
    - LM555 in subcircuits.
    - Wrong voltage in "PNP biasing" example.
    - Avoid program blocking in "End" GcBasic statement while debugging.

  RC2:
    - Switches not properly removed.
    - Wrong voltage in "PNP biasing" example final fix.
    - Avoid program blocking simulating AVR entering "sleep".
    - Wave gen. frequency accuracy.
    - Wrong Circuit Time.
    - Delete SR04 while running may crash.
    - GcBasic arrays errors in RamTable.
    - Crash setting mcu logic symbol=true.
    - Remove unintended components when removing single wire.
    - Crash removing Arduino in some cases (related to built-in led pin).
    - Problem in RamTable with variables changed in source code.

  RC3:
    - Plotter crashing randomly.
    - ResistorDip in Subcircuits.
    - Wire Lines disconnected in some cases.
    - Animate circuit resets after copy/paste.
    - Copy unintended components if previously selected.
    - Clock in subcircuits.
    - RamTable may crash when not debugging.
    - Filter file types in open dialogs.
    - Debugger fully interruptable.
    - AVR and PIC assemblers not found msg.
    - Adding led while circuit is running (wrong start step).
    - Some missing icons.
    - Relay: crash in changes when circuit running.
    - Keypad: crash in changes when circuit running.
    - LedMatrix: properties reset at size change.
    - 7segment: color reset at size change.
    - Subcircuit: error not cleared when package not found.
    - Volt. Reg: voltage not updated until circuit restart.
  
  SR1:
    - Current Source initialization after restart.
    - 555 CV pin wrong impedance.
    - 555 not cleanly removed.
    - Values with units in SubCircuits: wrong value.
    - Clock in SubCircuit not cleanly removed. 
    - Weird lines in connectors: Moving+Undo connectors with leght=0.
    - Uppercase asm instructions not recognized.
    - Crash compiling Pic asm with filenames containing spaces.
    - Import Circuit fails in some cases.
    - Pin Id changes not properly updated ( find pin by name fails ).
    - Bus: wrong bit0.
    - Editor: crash unindenting block with empty lines.
    - RAM/ROM wrong write cycle start in some cases.
    - I2C RAM: 1 byte address devices not working.
    - Simulation very slow with large circuits if animated.
    - Copy/Paste not working between simulide instances.
    
  SR2:
    - Crash starting new connector with circuit running and animated( 0.3.12-SR1 ).
    - Plotter randomly not working: Simulator update Counter not cleared.
    - Shapes not properly redrawn on size changes.
    - Id labels not well positioned in some components.
    - Wrong Cursor in New connector.
    - Closing Connector: end point bad position sometimes.
    
  SR3:
    - Crash hovering some external object over circuit canvas.
    - Circuit not updating properly when animated( 0.3.12-SR1 ).
    - Changes in Logic Devices properties while simulation running not updated.
    - Load Circuit: avoid connect to already connected pins.
    - Make 74HC and 74XX interchangeable.
    - BJT animation not updated.
    - BJT not working in vco example (0.3.12-RC2).
    - Fix some memory leaks.
    - Plotterwidget wrong scale in some cases.
    - 74HC4022 Logic Symbol not working.
    
  SR4:
    - Pic asm Compiler: gpasm does not find includes in project folder.
    - Time widget scrollBars shown in some systems.
    - Executable detected as shared lib in some builds.
    - AVR timer mode 14: Fast PWM, Top=ICR1, doesn't update in OCRX changes.
    - Send Text in Serial monitor limited to 50 characters.
    - ATtinyX4 wrong Pin asignation in Timer0,1.
    - ATtinyX4 wrong ADC definition.
    - Use Chip Symbol if Logic Symbol file not found.
    - Package should save file relative path, not absolute.
    - Create Subcircuit taking Component Circuit Id instead of unique Id.
    - Avoid dash "-" in Subcircuits Pin Ids.
    - Avoid save Backup in read only filesystems.
    - Subcircuits fixed (thanks to Sergey Roenko):

src/gui/circuitwidget/chip.cpp
/***************************************************************************
 *   Copyright (C) 2012 by santiago González                               *
 *   santigoro@gmail.com                                                   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, see <http://www.gnu.org/licenses/>.  *
 *                                                                         *
 ***************************************************************************/

#include <QDomDocument>

#include "chip.h"
#include "connector.h"
#include "circuit.h"
#include "utils.h"
#include "pin.h"
#include "simuapi_apppath.h"


Chip::Chip( QObject* parent, QString type, QString id )
    : Component( parent, type, id )
    , eElement( id.toStdString() )
{
    m_numpins = 0;
    m_isLS = false;
    
    m_pkgeFile = "";
    
    m_lsColor = QColor( 255, 255, 255 );
    m_icColor = QColor( 50, 50, 70 );
    
    setLabelPos( m_area.x(), m_area.y()-20, 0);
}
Chip::~Chip() {}

void Chip::initChip()
{
    //qDebug() << "Chip::initChip"<<m_pkgeFile;
    m_error = 0;
    
    QDir circuitDir = QFileInfo( Circuit::self()->getFileName() ).absoluteDir();
    QString fileNameAbs = circuitDir.absoluteFilePath( m_pkgeFile );

    QFile pfile( fileNameAbs );
    if( !pfile.exists() )   // Check if package file exist, if not try LS or no LS
    {
        if     ( m_pkgeFile.endsWith("_LS.package")) m_pkgeFile.replace( "_LS.package", ".package" );
        else if( m_pkgeFile.endsWith(".package"))    m_pkgeFile.replace( ".package", "_LS.package" );
        else qDebug() << "SubPackage::setPackage: No package files found.\nTODO: create dummy package\n";
        fileNameAbs = circuitDir.absoluteFilePath( m_pkgeFile );
    }
    QFile file( fileNameAbs );
    if( !file.open( QFile::ReadOnly | QFile::Text) )
    {
        MessageBoxNB( "Chip::initChip",
                  tr( "Cannot read file:\n%1:\n%2." ).arg(m_pkgeFile).arg(file.errorString()) );
          m_error = 1;
          return;
    }

    QDomDocument domDoc;
    if( !domDoc.setContent(&file) )
    {
         MessageBoxNB( "Chip::initChip",
                   tr( "Cannot set file:\n%1\nto DomDocument" ) .arg(m_pkgeFile));
         file.close();
         m_error = 2;
         return;
    }
    file.close();

    QDomElement root  = domDoc.documentElement();

    if( root.tagName()!="package" )
    {
        MessageBoxNB( "Chip::initChip",
                  tr( "Error reading Chip file:\n%1\nNo valid Chip" ) .arg(m_pkgeFile));
        m_error = 3;
        return;
    }

    m_width   = root.attribute( "width" ).toInt();
    m_height  = root.attribute( "height" ).toInt();
    m_numpins = root.attribute( "pins" ).toInt();
    
    foreach( Pin* pin, m_pin )
    {
        if( pin->connector() ) pin->connector()->remove();
        if( pin->scene() ) Circuit::self()->removeItem( pin );
        pin->reset();
        delete pin;
    }
    m_ePin.clear();
    m_pin.clear();
    m_ePin.resize( m_numpins );
    m_pin.resize( m_numpins );
    
    m_rigPin.clear();
    m_topPin.clear();
    m_lefPin.clear();
    m_botPin.clear();
    
    if( m_pkgeFile.endsWith( "_LS.package" )) m_isLS = true;
    else                                      m_isLS = false;

    if( m_isLS ) m_color = m_lsColor;
    else         m_color = m_icColor;

    m_area = QRect( 0, 0, 8*m_width, 8*m_height );
    //setTransformOriginPoint( togrid( boundingRect().center()) );
    
    setShowId( true );

    QDomNode node = root.firstChild();

    int chipPos = 0;

    while( !node.isNull() )
    {
        QDomElement element = node.toElement();
        if( element.tagName() == "pin" )
        {
            QString type  = element.attribute( "type" );
            QString label = element.attribute( "label" );
            QString id    = element.attribute( "id" );
            QString side  = element.attribute( "side" );
            int     pos   = element.attribute( "pos" ).toInt();

            int xpos = 0;
            int ypos = 0;
            int angle = 0;

            if( side=="left" )
            {
                xpos = -8;
                ypos = 8*pos;
                angle = 180;
            }
            else if( side=="top")
            {
                xpos = 8*pos;
                ypos = -8;
                angle = 90;
            }
            else if( side=="right" )
            {
                xpos =  m_width*8+8;
                ypos = 8*pos;
                angle = 0;
            }
            else if( side=="bottom" )
            {
                xpos = 8*pos;
                ypos =  m_height*8+8;
                angle = 270;
            }
            chipPos++;              
            addPin( id, type, label, chipPos, xpos, ypos, angle );
        }
        node = node.nextSibling();
    }
}

void Chip::addPin( QString id, QString type, QString label, int pos, int xpos, int ypos, int angle )
{
    Pin* pin = new Pin( angle, QPoint(xpos, ypos), m_id+"-"+id, pos-1, this ); // pos in package starts at 1
    
    //m_pinMap[id] = pin;
    
    pin->setLabelText( label );
    
    if     ( type == "inverted" ) pin->setInverted( true );
    else if( type == "unused" )   pin->setUnused( true );
    else if( type == "null" )
    {
        pin->setVisible( false );
        pin->setLabelText( "" );
    }
    if     ( angle == 0 )   m_rigPin.append( pin );
    else if( angle == 90 )  m_topPin.append( pin );
    else if( angle == 180 ) m_lefPin.append( pin );
    else if( angle == 270 ) m_botPin.append( pin );

    if( m_isLS ) pin->setLabelColor( QColor( 0, 0, 0 ) );

    m_ePin[pos-1] = pin;
    m_pin[pos-1]  = pin;
}

/*void Chip::updatePin( QString id, QString type, QString label, int pos, int xpos, int ypos, int angle )
{
    Pin* pin = m_pin[pos-1]; // pos in package starts at 1

    pin->setLabelText( label );
    pin->setPos( QPoint(xpos, ypos) );
    
    int oldAngle = pin->pinAngle();
    if( angle != oldAngle )
    {
        if     ( oldAngle == 0 )   m_rigPin.removeOne( pin );
        else if( oldAngle == 90 )  m_topPin.removeOne( pin );
        else if( oldAngle == 180 ) m_lefPin.removeOne( pin );
        else if( oldAngle == 270 ) m_botPin.removeOne( pin );
        
        if     ( angle == 0 )   m_rigPin.append( pin );
        else if( angle == 90 )  m_topPin.append( pin );
        else if( angle == 180 ) m_lefPin.append( pin );
        else if( angle == 270 ) m_botPin.append( pin );
    }
    
    pin->setPinAngle( angle );
    pin->setLabelPos();
    
    if( type == "inverted" ) pin->setInverted( true );
    else                     pin->setInverted( false );
    
    if( type == "unused" )   pin->setUnused( true );
    else                     pin->setUnused( false );

    if( type == "null" )
    {
        pin->setVisible( false );
        pin->setLabelText( "" );
    }
    else pin->setVisible( true );
   
    if( m_isLS ) pin->setLabelColor( QColor( 0, 0, 0 ) );
    else                  pin->setLabelColor( QColor( 250, 250, 200 ) );
    
    pin->isMoved();
}*/

bool Chip::logicSymbol()
{
    return m_isLS;
}

void Chip::setLogicSymbol( bool ls )
{
    if( m_isLS == ls ) return;
    
    if     ( m_pkgeFile.endsWith("_LS.package")) m_pkgeFile.replace( "_LS.package", ".package" );
    else if( m_pkgeFile.endsWith(".package"))    m_pkgeFile.replace( ".package", "_LS.package" );

    m_error = 0;
    Chip::initChip();
    
    if( m_error == 0 )  Circuit::self()->update();
}

void Chip::remove()
{
    /*for( uint i=0; i<m_ePin.size(); i++ )
    {
        Pin* pin = static_cast<Pin*>(m_ePin[i]);
        if( pin->connector() ) pin->connector()->remove();
    }*/
    Component::remove();
}

void Chip::contextMenuEvent( QGraphicsSceneContextMenuEvent* event )
{
    event->accept();
    QMenu *menu = new QMenu();
    /*QAction *loadAction = menu->addAction( QIcon(":/fileopen.png"),tr("Load firmware") );
    connect( loadAction, SIGNAL(triggered()), this, SLOT(slotLoad()) );

    QAction *reloadAction = menu->addAction( QIcon(":/fileopen.png"),tr("Reload firmware") );
    connect( reloadAction, SIGNAL(triggered()), this, SLOT(slotReload()) );*/

    menu->addSeparator();

    Component::contextMenu( event, menu );
    menu->deleteLater();
}

void Chip::paint( QPainter *p, const QStyleOptionGraphicsItem *option, QWidget *widget )
{
    Component::paint( p, option, widget );

    p->drawRoundedRect( m_area, 1, 1);

    if( !m_isLS )
    {
        p->setPen( QColor( 170, 170, 150 ) );
        p->drawArc( boundingRect().width()/2-6, -4, 8, 8, 0, -2880 /* -16*180 */ );
    }
}

#include "moc_chip.cpp"

